crates/client/data-availability/src/lib.rs

```diff
@@ -0,0 +1,159 @@
+pub enum DaLayer {
+    Celestia,
+    Ethereum,
+    Avail,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Deserialize, Default)]
+pub enum DaMode {
+    #[serde(rename = "validity")]
+    Validity,
+    #[serde(rename = "volition")]
+    Volition,
+    #[serde(rename = "validium")]
+    #[default]
+    Validium,
+}
+
+#[async_trait]
+pub trait DaClient: Send + Sync {
+    fn get_mode(&self) -> DaMode;
+    async fn last_published_state(&self) -> Result<I256>;
+    async fn publish_state_diff(&self, state_diff: Vec<U256>) -> Result<()>;
+}
+
+impl<B, C> DataAvailabilityWorker<B, C>
+where
+    B: BlockT,
+    C: ProvideRuntimeApi<B>,
+    C: BlockchainEvents<B> + 'static,
+{
+    pub async fn prove_current_block(da_mode: DaMode, client: Arc<C>, madara_backend: Arc<mc_db::Backend<B>>) {
+        let mut storage_event_st = client.storage_changes_notification_stream(None, None).unwrap();
+
+        while let Some(storage_event) = storage_event_st.next().await {
+            // Locate and encode the storage change
+            let mut _deployed_contracts: Vec<String> = Vec::new();
+            let mut nonces: HashMap<&[u8], &[u8]> = HashMap::new();
+            let mut storage_diffs: HashMap<&[u8], StorageWrites> = HashMap::new();
+
+            // Locate and encode the storage change
+            for event in storage_event.changes.iter() {
+                let mut prefix = event.1.0.as_slice();
+                let mut key: &[u8] = &[];
+                if prefix.len() > 32 {
+                    let raw_split = prefix.split_at(32);
+                    prefix = raw_split.0;
+                    key = raw_split.1;
+                }
+
+                if prefix == *utils::SN_NONCE_PREFIX {
+                    if let Some(data) = event.2 {
+                        nonces.insert(key, data.0.as_slice());
+                    }
+                }
+
+                if prefix == *utils::SN_STORAGE_PREFIX {
+                    if let Some(data) = event.2 {
+                        // first 32 bytes = contract address, second 32 bytes = storage variable
+                        let write_split = key.split_at(32);
+
+                        storage_diffs
+                            .entry(write_split.0)
+                            .and_modify(|v| v.push((write_split.1, data.0.as_slice())))
+                            .or_insert(vec![(write_split.1, data.0.as_slice())]);
+                    }
+                }
+            }
+
+            let state_diff = utils::pre_0_11_0_state_diff(storage_diffs, nonces);
+
+            // Store the DA output from the SN OS
+            if let Err(db_err) = madara_backend.da().store_state_diff(&storage_event.block, state_diff) {
+                log::error!("db err: {db_err}");
+            };
+
+            match da_mode {
+                DaMode::Validity => {
+                    // Submit the StarkNet OS PIE
+                    if let Ok(job_resp) = sharp::submit_pie("test") {
```

Comment:
Is it just a placeholder until further development?
Or should it be the app-chain responsibility to edit this code?

Reply:
// TODO: Validity Impl
                    // run the Starknet OS with the Cairo VM
                    // extract the PIE from the Cairo VM run
                    // pass the PIE to `submit_pie` and zip/base64 internal
  
  added comment